**Data Preparations**:
Download the dataset titanic.xls in your local computer and load it into a DataFrame named titanic_df. Display the first 10 rows of dataset to understand the data structure.

In [ ]:
import pandas as pd

# Load the dataset into a Pandas DataFrame
titanic_df = pd.read_excel('/content/sample_data/titanic (1).xls')

# Display the first 10 rows of the dataset
print(titanic_df.head(10))


   pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   
5       1         1                              Anderson, Mr. Harry    male   
6       1         1                Andrews, Miss. Kornelia Theodosia  female   
7       1         0                           Andrews, Mr. Thomas Jr    male   
8       1         1    Appleton, Mrs. Edward Dale (Charlotte Lamson)  female   
9       1         0                          Artagaveytia, Mr. Ramon    male   

       age  sibsp  parch    ticket      fare    cabin embarked boat   body  \
0  29.0000      0      0     24160  211.3

Create two DataFrames from titanic_df: one containing PassengerId, Name, and Pclass, and the other containing PassengerId, Survived, and Age.

Task: Perform an inner join on these two DataFrames using PassengerId as the key.

In [ ]:
titanic_df['PassengerID'] = titanic_df.reset_index().index
print(titanic_df.head(10))


   pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   
5       1         1                              Anderson, Mr. Harry    male   
6       1         1                Andrews, Miss. Kornelia Theodosia  female   
7       1         0                           Andrews, Mr. Thomas Jr    male   
8       1         1    Appleton, Mrs. Edward Dale (Charlotte Lamson)  female   
9       1         0                          Artagaveytia, Mr. Ramon    male   

       age  sibsp  parch    ticket      fare    cabin embarked boat   body  \
0  29.0000      0      0     24160  211.3

In [ ]:

# Create the first DataFrame containing PassengerId, Name, and Pclass
df1 = titanic_df[['PassengerID', 'name', 'pclass']]

# Create the second DataFrame containing PassengerId, Survived, and Age
df2 = titanic_df[['PassengerID', 'survived', 'age']]

# Perform an inner join on the two DataFrames using PassengerId as the key
merged_df = pd.merge(df1, df2, on='PassengerID', how='inner')

# Display the first few rows of the merged DataFrame to verify
print(merged_df)

      PassengerID                                             name  pclass  \
0               0                    Allen, Miss. Elisabeth Walton       1   
1               1                   Allison, Master. Hudson Trevor       1   
2               2                     Allison, Miss. Helen Loraine       1   
3               3             Allison, Mr. Hudson Joshua Creighton       1   
4               4  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)       1   
...           ...                                              ...     ...   
1304         1304                             Zabour, Miss. Hileni       3   
1305         1305                            Zabour, Miss. Thamine       3   
1306         1306                        Zakarian, Mr. Mapriededer       3   
1307         1307                              Zakarian, Mr. Ortin       3   
1308         1308                               Zimmerman, Mr. Leo       3   

      survived      age  
0            1  29.0000  
1          

Outer Join: (20 points)
Perform an outer join on the previous two DataFrames.

In [ ]:
merged_df_outer = pd.merge(df1, df2, on='PassengerID', how='outer')

# Display the first few rows of the merged DataFrame to verify the join
print(merged_df_outer)

      PassengerID                                             name  pclass  \
0               0                    Allen, Miss. Elisabeth Walton       1   
1               1                   Allison, Master. Hudson Trevor       1   
2               2                     Allison, Miss. Helen Loraine       1   
3               3             Allison, Mr. Hudson Joshua Creighton       1   
4               4  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)       1   
...           ...                                              ...     ...   
1304         1304                             Zabour, Miss. Hileni       3   
1305         1305                            Zabour, Miss. Thamine       3   
1306         1306                        Zakarian, Mr. Mapriededer       3   
1307         1307                              Zakarian, Mr. Ortin       3   
1308         1308                               Zimmerman, Mr. Leo       3   

      survived      age  
0            1  29.0000  
1          

Discuss the differences in the results compared to the inner join.
There is no difference between inner join and outer join.

**GroupBy Exercise:**
Use the original titanic_df.

Task 1: Group by Pclass and calculate the average Fare and Age for each class.
Task 2: Group by Embarked and Sex, and find the count of passengers for each combination.

In [ ]:
# Task 1
grouped_by_pclass = titanic_df.groupby('pclass').agg({'fare': 'mean', 'age': 'mean'})

# Display the result
print("Average Fare and Age by Pclass:")
print(grouped_by_pclass)


Average Fare and Age by Pclass:
             fare        age
pclass                      
1       87.508992  39.159918
2       21.179196  29.506705
3       13.302889  24.816367


In [ ]:
# Task 2
grouped_by_embarked_sex = titanic_df.groupby(['embarked', 'sex']).size().reset_index(name='Count')

# Display the result
print("Count of Passengers by Embarked and Sex:")
print(grouped_by_embarked_sex)


Count of Passengers by Embarked and Sex:
  embarked     sex  Count
0        C  female    113
1        C    male    157
2        Q  female     60
3        Q    male     63
4        S  female    291
5        S    male    623


**Explore More**:
Task 1. Use the groupby method to calculate the proportion of passengers that survived by sex.
Task 2. Create age categories: children (under 14 years), adolescents (14-20), adult (21-64), and senior(65+), and calculate survival proportions by age category, pclass and sex.

In [ ]:
# Task 1
survival_by_sex = titanic_df.groupby('sex')['survived'].mean()

# Display the result
print("Survival Proportion by Sex:")
print(survival_by_sex)


Survival Proportion by Sex:
sex
female    0.727468
male      0.190985
Name: survived, dtype: float64


In [ ]:
import numpy as np

# Define a function to categorize age
def categorize_age(age):
    if pd.isnull(age):
        return 'Unknown'  # Optional: handle missing values
    elif age < 14:
        return 'Children'
    elif 14 <= age <= 20:
        return 'Adolescents'
    elif 21 <= age <= 64:
        return 'Adult'
    else:
        return 'Senior'

# Create a new column 'AgeCategory' using the defined function
titanic_df['AgeCategory'] = titanic_df['age'].apply(categorize_age)

# Group by 'AgeCategory', 'Pclass', and 'Sex', and calculate survival proportions
survival_by_age_pclass_sex = titanic_df.groupby(['AgeCategory', 'pclass', 'sex'])['survived'].mean().unstack()

# Display the result
print("Survival Proportions by Age Category, Pclass, and Sex:")
print(survival_by_age_pclass_sex)


Survival Proportions by Age Category, Pclass, and Sex:
sex                   female      male
AgeCategory pclass                    
Adolescents 1       1.000000  0.200000
            2       0.923077  0.117647
            3       0.542857  0.125000
Adult       1       0.965517  0.343284
            2       0.868421  0.078125
            3       0.441860  0.159836
Children    1       0.000000  1.000000
            2       1.000000  1.000000
            3       0.483871  0.324324
Senior      1       1.000000  0.142857
            2            NaN  0.000000
            3            NaN  0.000000
Unknown     1       1.000000  0.285714
            2       0.666667  0.153846
            3       0.531250  0.111111


**Analysis**:
**Gender Disparity**: Women generally had higher survival rates across all age categories and classes compared to men, highlighting the gender disparity in survival likelihood. For instance, adult women in 1st class had a survival rate of approximately 96.55%, while their male counterparts had a rate of 34.33%. This trend persists across all classes and age groups, underlining the "women and children first" policy during the evacuation.

**Class Influence**: The class appears to significantly impact survival rates, with higher classes showing markedly higher survival rates. Notably, adolescents and adults in the 1st class exhibit the highest survival rates among their respective age groups, emphasizing the advantage of being in a higher socio-economic class during the disaster.

**Children's Survival Rates**: The survival rates for children varied significantly across classes. Children in the 2nd class had a survival rate of 100% for both females and males, which is the highest survival rate observed in the dataset. However, children in the 3rd class had lower survival rates, with 48.39% for females and 32.43% for males, suggesting that class disparities also affected the youngest passengers.

**Senior Passengers**: Senior passengers had very low representation, making their survival rates less statistically robust. However, it is noteworthy that senior females in the 1st class had a survival rate of 100%, whereas senior males had much lower rates across all classes.

**Adolescents**: Adolescent females in the 1st and 2nd classes had high survival rates (100% and 92.31%, respectively), far exceeding those of their male counterparts. This further illustrates the advantage of being female and in a higher class.

**Unknown Age Group**: The survival rates for passengers whose age was unknown also reflected the general trend of higher survival rates for females. However, due to the uncertainty about the age distribution within this group, drawing specific conclusions is challenging.

**Conclusion**:
The analysis underscores the significant influence of gender and class on survival rates aboard the Titanic. Women and children, particularly those in higher classes, had markedly higher survival rates, reflecting the prioritization of their evacuation. Class disparities were evident across all age groups, with 1st and 2nd-class passengers enjoying considerably higher chances of survival than those in the 3rd class. While the dataset provides valuable insights into the tragic event, it also highlights the socio-economic inequalities that played a role in survival outcomes.